test/functional/apm_logging_tests.js

```diff
@@ -0,0 +1,225 @@
+      expect(data.numLines).to.equal(2);
+      expect(data.sdam).to.be.empty;
+      done(err);
+    };
+    runTestClient(this, '?monitor=command&monitorOut=apm_test_log', cb);
+  });
+
+  // skip for now: this test depends on the new fixes in the core uri-parser
+  it.skip('log both command and sdam events using `monitor=command,sdam`', function(done) {
+    const cb = err => {
+      const logFile = fs.readFileSync(testLogFile, 'utf8');
+      const data = getFileData(logFile);
+
+      expect(data.numLines).to.equal(8);
+      expect(data.sdam).to.not.be.empty;
+      expect(data.command).to.not.be.empty;
+      done(err);
+    };
+    runTestClient(this, '?monitor=command,sdam&monitorOut=apm_test_log', cb);
+  });
+
+  it('log both command and sdam events using `monitor=all`', function(done) {
+    const cb = err => {
+      const logFile = fs.readFileSync(testLogFile, 'utf8');
+      const data = getFileData(logFile);
+
+      expect(data.numLines).to.equal(8);
+      expect(data.sdam).to.not.be.empty;
+      expect(data.command).to.not.be.empty;
+      done(err);
+    };
+    runTestClient(this, '?monitor=all&monitorOut=apm_test_log', cb);
+  });
+
+  it('should log correct event objects in the same order they were emitted', function(done) {
+    const cb = (err, events) => {
+      const logFile = fs.readFileSync(testLogFile, 'utf8');
+      const data = getFileData(logFile);
+
+      expect(events).to.deep.equal(data.eventStrings);
+      done(err);
+    };
+    runTestClient(this, '?monitor=all&monitorOut=apm_test_log', cb);
+  });
+
+  it('should call callback with error when log filename cannot be resolved', {
+    metadata: { requires: { node: '<10.0.0' } },
+    test: function(done) {
+      const configuration = this.configuration;
+      const url = configuration.url() + '?monitor=all&monitorOut=invalid:name';
+      const client = new MongoClient(url, { w: 1, useNewUrlParser: true });
+
+      client.connect(function(err) {
+        expect(err).to.exist;
+        console.log(err);
+        expect(err.name).to.equal('TypeError');
+        expect(err.message).to.equal(`Path must be a string. Received { invalid: 'name' }`);
+        done();
+      });
+    }
+  });
+
+  it('should call callback with error when log filename cannot be resolved', {
+    metadata: { requires: { node: '>=10.0.0' } },
+    test: function(done) {
+      const configuration = this.configuration;
+      const url = configuration.url() + '?monitor=all&monitorOut=invalid:name';
+      const client = new MongoClient(url, { w: 1, useNewUrlParser: true });
+
+      client.connect(function(err) {
+        expect(err).to.exist;
+        expect(err.name).to.equal('TypeError [ERR_INVALID_ARG_TYPE]');
+        expect(err.message).to.equal(
+          `The "path" argument must be of type string. Received type object`
+        );
+        done();
+      });
+    }
+  });
+
+  // TODO: how to test throwing errors if we don't know exactly when the error will be thrown?
```

Comment:
Not sure how to test for stream errors that happen at some undefined time, potentially after the client has already connected - for example, in this test, trying to connect with the custom filepath `nonexistent/file/path` causes the writable output stream to emit an `error` event at some arbitrary time, which can't be caught by a `try/catch` or anything of that sort

Reply:
This might be a case where testing with a real file could be useful. You could specify a temp file, and then delete it after client initialization thus triggering the error handler